### PR TITLE
feat(agui): expose AG-UI protocol via /protocol/agui/chat endpoint

### PR DESCRIPTION
**Pull request by mautops**
## Description

AgentScope 2.0 ships an `AGUIProtocolMiddleware`, but QwenPaw does not expose the AG-UI protocol to external consumers. This PR adds a dedicated `/protocol/agui/chat` endpoint that streams agent responses as standard AG-UI protocol events over SSE, fully isolated from existing services (`/console/chat` is untouched).

**Motivation:** Exposing AG-UI lets any AG-UI-compatible client (copilots, orchestration layers, dashboards) consume QwenPaw agents through a standardized event protocol, complementing the existing console/ACP channels.

### Architecture

```
src/qwenpaw/app/protocol/
├── __init__.py
└── agui/
    ├── __init__.py          # re-exports router
    └── router.py            # /protocol/agui/chat (SSE passthrough)
```

- `router.py` normalizes the body to a real `AgentRequest`, streams the agent lifecycle through the runtime, and converts each native AgentScope `AgentEvent` to AG-UI via AgentScope 2.0's built-in `AGUIProtocolMiddleware`, framing the result as SSE (`data: <json>\n\n`).
- No custom converter — the endpoint exports AgentScope's native runtime envelope format directly.
- **Runtime unification:** the request lifecycle lives in a single `Runtime._lifecycle()` core. `run()` projects events through the SSE `Envelope`; `run_agent_events()` yields raw `AgentEvent`s for protocol adapters. Slash commands and hook short-circuits are projected into both streams (as synthetic `AgentEvent`s for the raw path), so `/console/chat` and `/protocol/agui/chat` behave consistently.
- **Dependency:** `ag-ui-protocol` is an explicit runtime dependency (importable as `ag_ui`); its presence is validated before the SSE response starts.

### Security Considerations

The endpoint reuses the existing request auth path (`get_agent_for_request`) and inherits the same workspace/agent scoping as `/console/chat`. No new credentials or network egress.

### Evidence

Tested end-to-end locally against a running QwenPaw instance. Normal agent turns stream the full `RUN_STARTED → text message events → RUN_FINISHED` lifecycle; slash commands (e.g. `/version`) dispatch through the slash-command registry and stream their result; conversion failures emit `RUN_ERROR` and terminate the stream rather than dropping lifecycle-critical events. SSE heartbeats keep long-running requests alive.

No changes to `/console/chat` — existing endpoint continues to stream as before (byte-identical envelope output verified).

### Checklist

- [x] I ran `pre-commit run --all-files` locally and it passes
- [x] `black` + `flake8` pass on the changed files
- [x] I ran tests locally (`pytest` or as relevant) and they pass
- [ ] Documentation updated (if needed)
- [x] Ready for review

### CODE REVIEW

**Comment by Leirunlin** [comment]
@mautops Hi, thanks for the PR.
 - How about reconsidering the integration boundary and reuse AgentScope 2.0’s existing AG-UI protocol conversion where possible? 
 - Ideally, this route should consume the raw AgentEvent stream instead of introducing a separate QwenPaw Envelope-to-AG-UI conversion layer. Maybe:
 ```
                         +-> QwenPaw Envelope -> existing QwenPaw endpoints
 AgentScope AgentEvent --|
                         +-> AgentScope AG-UI adapter -> AG-UI endpoint
 ```
We can discuss more about the design.

 And:
 - Please remove the unrelated .gitignore change.
 - Keep comments and docstrings concise and consistent, instead of mixing Chinese/English comments and some irrelevant notes.
 - Check the existing AI review comments.
 - ensure all pre-commit checks pass before requesting another review.

**Comment by Leirunlin** [comment]
@mautops Thanks for the update. The direction is now correct, but a few issues remain:

1. Use an explicit AgentEvent API
Please consider `workspace.stream_agent_events(request)` instead of propagating `raw=True` through Workspace, Runtime, and Executor.
`stream_query()` currently promises QwenPaw Envelopes, while raw mode changes its return type and behavior. Slash commands and hook short circuits still call `envelope.from_msg()`, so the “raw” stream can mix AgentEvent and Envelope objects. The broader Msg/event design can be handled separately, but this endpoint should receive a type-consistent stream.

2. Handle conversion failures
Do not skip failed events. Emit RUN_ERROR and terminate the stream; otherwise lifecycle-critical events may be lost.
3. Restore the dependency
ag-ui-protocol is an optional AgentScope dependency and is not installed by default. Add it explicitly and validate it before starting the SSE response.

4. Small issues
      - Import AGUIProtocolMiddleware from the public agentscope.app.middleware API.
      - Preserve SSE heartbeats for long-running requests. (Minor and Optional, can be omit from this PR)
      - Clarify that the endpoint currently accepts QwenPaw AgentRequest, not AG-UI RunAgentInput. (Minor and Optional, can be omit from this PR)

**Comment by Leirunlin** [comment]
@mautops  Thanks for the update. The overall direction looks good. I tested the latest commit locally and confirmed that the normal agent path works after fixing the dependency import, and both the conversion-error handling and SSE heartbeat behavior are working as expected.

### Blocking issues

1. **Incorrect dependency import**  
   `ag-ui-protocol` exposes `ag_ui`, not `ag_ui_protocol`.  
   ```python
   import ag_ui           # works
   import ag_ui_protocol  # ModuleNotFoundError
   ```  
   `/protocol/agui/chat` currently returns HTTP 500. Changing the validation import to `import ag_ui` resolves the issue.

2. **Slash commands bypass QwenPaw dispatch**  
   I tested `/version` against both endpoints:  
   - `/api/console/chat` returned the daemon version directly.  
   - `/protocol/agui/chat` invoked the model, executed `execute_shell_command`, and then called the model again.  

   `run_agent_events()` currently bypasses the slash-command registry, which changes existing QwenPaw behavior and should be addressed.

3. **Runtime lifecycle is duplicated**  
   `run_agent_events()` copies most of `Runtime.run()` while altering command, hook, and cancellation behavior. This introduces two lifecycle implementations that can diverge over time. Please unify orchestration and vary only the output projection.

### Additional issues

- **Hook `SHORT_CIRCUIT` payloads are dropped**  
  This results in an empty AG-UI stream. It mainly affects plugin-defined hooks, so it's less urgent than the slash-command handling, but the behavior should either be properly supported or explicitly rejected.
I have not tested it, you can have it double checked.

- **The minimal `Envelope` in `run_agent_events()` seems that it never consumes raw events**  
  As a result, it cannot collect partial text or tool state during cancellation, despite what the comment claims.
  I have not tested it, you can review and check it again.

- Please update the PR description to reflect the current state of the changes.

- Please keep comments clean and accurate.

**Comment by mautops** [comment]
@Leirunlin Thanks for the detailed review. All three blocking issues plus the additional points are addressed in `479fd874`:

**1. Dependency import** — Fixed. The validation now uses `import ag_ui` (the module exposed by `ag-ui-protocol`), and `AGUIProtocolMiddleware` is imported from the public `agentscope.app.middleware` API. Verified locally that `/protocol/agui/chat` no longer returns HTTP 500.

**2. Slash commands bypass dispatch** — Fixed. The slash-command registry now runs in the agent-events path. The request lifecycle was unified into a single `Runtime._lifecycle()` core shared by `run()` and `run_agent_events()`, with only the output projection varying. Slash-command and hook `SHORT_CIRCUIT` results are projected as synthetic `AgentEvent`s (`ReplyStart → TextBlock* → ReplyEnd`), so the stream stays type-consistent and `/version` returns the daemon version directly instead of invoking the model. Verified end-to-end: `/version` streams `RUN_STARTED → TEXT_MESSAGE_CONTENT("qwenpaw 2.0") → RUN_FINISHED`.

**3. Runtime lifecycle duplicated** — Fixed. One orchestration core (`_lifecycle`), two thin projections. `run()` maps events through the SSE `Envelope` (byte-identical output verified against the old `run()`); `run_agent_events()` yields raw `AgentEvent`s. Error handling (ON_ERROR → cancel-save → cancel/error envelope) and FINALLY cleanup are shared, so they cannot diverge.

**Additional points:**
- **SHORT_CIRCUIT payloads dropped** — Fixed. They now flow through `_project_msg` and produce a proper stream (empty-stream case eliminated).
- **Minimal Envelope never consumed raw events** — Removed the fake `Envelope`. `run_agent_events()` no longer sets `ctx._envelope`; cancel-save persists `agent.state_dict()` honestly, and partial-stream injection correctly no-ops for the raw path (there is no envelope state machine to reconstruct partial text from).
- **PR description** — Updated to reflect the current architecture.
- **Comments** — Cleaned; no mixed-language or stale notes.
- **Pre-commit** — All hooks pass on the changed files.

One design note: the `AGUIProtocolMiddleware` converter is instantiated per request, so its `_tool_result_buffers` / `_last_model_name` state is never shared across concurrent SSE connections.

File: src/qwenpaw/app/protocol/agui/router.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    FastAPI may deliver the body as a dict (when the JSON does not fully
+    match the schema) or as a validated AgentRequest. We coerce both to a
+    real AgentRequest so that ``workspace.stream_query`` receives a request
+    whose ``input`` field actually carries the user's message — the native
+    ``{content_parts, ...}`` dict shape is channel-internal and must NOT be
+    passed to the runtime directly (its ``input`` would default to empty).
+    """
+    if isinstance(request_data, AgentRequest):
+        return request_data
+    return AgentRequest.model_validate(request_data)
```
[the remark this answers is not in the document]
Comment (mautops) [reply src/qwenpaw/app/protocol/agui/router.py]: Addressed in 479fd874. The current _normalize_request() catches ValidationError and raises HTTP 422, and rejects empty input with 422.

File: src/qwenpaw/app/protocol/agui/converter.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+            return AGUICustomEvent(
+                name="content_unknown",
+                value=event,
+            )
+
+        # Check for tool events
+        if event.get("object") == "tool_call":
+            tool_call_id = event.get("id", "")
+            tool_name = event.get("name", "")
+            status = event.get("status", "")
```
[the remark this answers is not in the document]
Comment (mautops) [reply src/qwenpaw/app/protocol/agui/converter.py]: Obsolete — this comment referenced the custom converter (src/qwenpaw/app/protocol/agui/converter.py), which was removed in a later revision. The endpoint now converts native AgentScope AgentEvents through AgentScope's built-in AGUIProtocolMiddleware, which emits the standard TOOL_CALL_* events directly.

File: src/qwenpaw/app/protocol/agui/converter.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                name="message_unknown",
+                value=event,
+            )
+
+        # Check for content events (nested inside messages)
+        if event.get("object") == "content":
+            content_type = event.get("type", "")
+            msg_id = event.get("msg_id", "")
+            text = event.get("text", "")
+
```
[the remark this answers is not in the document]
Comment (mautops) [reply src/qwenpaw/app/protocol/agui/converter.py]: Obsolete — this comment referenced the removed custom converter. The current endpoint streams native AgentScope AgentEvents through AGUIProtocolMiddleware, so delta/final suppression is handled by AgentScope's own event stream (TextBlockDeltaEvent vs TextBlockEndEvent), not the envelope's delta flag.

File: src/qwenpaw/app/protocol/agui/converter.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        if isinstance(content, str):
+            return content
+        return json.dumps(content, ensure_ascii=False)
+
+
+def create_converter() -> QwenPawToAGUIConverter:
+    """创建新的转换器实例.
+
+    用于每个请求创建独立的转换器，避免并发状态冲突。
+
```
[the remark this answers is not in the document]
Comment (mautops) [reply src/qwenpaw/app/protocol/agui/converter.py]: Obsolete — this comment referenced the removed custom converter. The remaining conversion logic (AGUIProtocolMiddleware) is AgentScope-owned and tested upstream; QwenPaw's own runtime paths are covered by the unit suite in tests/unit/runtime/. Added an end-to-end verification of the SSE stream in the latest revision.

**Comment by Leirunlin** [comment]
LGTM
